server/src/main/java/org/elasticsearch/action/bulk/BulkPrimaryExecutionContext.java

```diff
@@ -191,12 +193,38 @@ public void resetForMappingUpdateRetry() {
         resetForExecutionRetry();
     }
 
+    /**
+     * Don't bother the master node if the mapping update is a noop.
+     * This may happen if there was a concurrent mapping update that added the same field.
+     *
+     * @param mappingVersion the current mapping version. This is used to guard against infinite loops.
+     * @throws IllegalStateException if retried multiple times with the same mapping version, to guard against infinite loops.
+     */
+    public void resetForNoopMappingUpdateRetry(long mappingVersion) {
+        assert assertInvariants(ItemProcessingState.TRANSLATED);
+        if (noopMappingUpdateRetryForMappingVersion == mappingVersion) {
+            // seems like we're in a live lock/infinite loop here
+            // we've already re-tried and are about to retry again
+            // as no state has changed in the meantime (the mapping version is still the same),
+            // we can't expect another retry would yield a different result
+            // a possible cause:
+            // maybe we added more dynamic mappers in DocumentParserContext.addDynamicMapper than possible according to the field limit
+            // the additional fields are then ignored by the mapping merge and the process repeats
+            throw new IllegalStateException(
```

Comment:
Can we assert false here too? That would satisfy my worries about this - we should consider it a bug. I am ok to throw in that case (though the client is then likely to retry).